### PR TITLE
8291945: Add OSInfo API for static OS information

### PR DESCRIPTION
**Pull request by iklam**
As suggested by @tstuefe in the review of #9600, many of the `os` APIs simply return a value that was initialized during VM start up. E.g., `os::vm_page_size()`. I have added a new class `OSInfo` for managing such info

- Simplifies the per-OS implementation
- Make the "init-only-once" check consistent
- In some cases, the simpler `OSInfo` class can be used in lieu of the more complex `os` class. This improves HotSpot build time (e.g., relocInfo.hpp).

Note: I moved only 2 fields in this PR as a first step. If this works well I plan to move more fields in the future.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291945](https://bugs.openjdk.org/browse/JDK-8291945): Add OSInfo API for static OS information

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [9e918e8e](https://git.openjdk.org/jdk/pull/9758/files/9e918e8eff9a1d6272e3e2b67a649870f266ab3f)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [3da77fcb](https://git.openjdk.org/jdk/pull/9758/files/3da77fcb7650c870001815739b5290c74dd53474)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9758/head:pull/9758` \
`$ git checkout pull/9758`

Update a local copy of the PR: \
`$ git checkout pull/9758` \
`$ git pull https://git.openjdk.org/jdk pull/9758/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9758`

View PR using the GUI difftool: \
`$ git pr show -t 9758`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9758.diff">https://git.openjdk.org/jdk/pull/9758.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/osInfo.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ */
+
+#ifndef SHARE_RUNTIME_OSINFO_HPP
+#define SHARE_RUNTIME_OSINFO_HPP
+
+#include "memory/allStatic.hpp"
+#include "utilities/debug.hpp"
+
+// Static information about the operation system. Initialized exactly once
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/osInfo.hpp:31]: s/operation/operating

File: src/hotspot/share/runtime/stackValue.hpp [hunk 2]
```diff
@@ -29,6 +29,7 @@
 #include "runtime/handles.hpp"
 
 class BasicLock;
+class frame;
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/stackValue.hpp:32]: Seems unrelated to this change. ?/

**Review by dholmes-ora — COMMENTED** [review]
Seems reasonable, but I initially expected to see the os:: API removed and replaced by the OSInfo:: API, rather than OSInfo just being an internal implementation detail.

File: src/hotspot/share/runtime/stackValue.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/runtime/stackValue.hpp:32]: This file was indirectly including frame.hpp via os.hpp. Since I removed os.hpp from this file, I needed to add a declaration for the `frame` class.

**Comment by iklam** [comment]
> Seems reasonable, but I initially expected to see the os:: API removed and replaced by the OSInfo:: API, rather than OSInfo just being an internal implementation detail.

I left os::vm_page_size() because:

- It's used by a lot of places. I don't want this PR to be overwhelmed by such changes.
- os::vm_page_size() is often used together with other os::xxx functions related to memory allocation.

So at this point, OSInfo is a just lower-level API if you just want to find out the value of some OS parameters. So far I've used it in two popular headers to avoid including os.hpp

File: src/hotspot/share/runtime/stackValue.hpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/stackValue.hpp:32]: I don't see any such removal in this file. ??

**Comment by dholmes-ora** [comment]
Okay I think it would be good to switch to the OSInfo API in the future, so that we reduce the scope of os even further and more clearly delineate an API to request services from the OS, and a simple query API for the OS.

**Review by dholmes-ora — APPROVED** [review]

File: src/hotspot/share/runtime/stackValue.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/runtime/stackValue.hpp:32]: Oh I was mistaken. What I did was removing os.hpp from relocInfo.hpp and safepointMechanismhpp. This exposed the fact that stackValue.hpp was using the `frame` class without a proper declaration.

**Review by tstuefe — APPROVED** [review]
Looks good!

Cheers, Thomas

**Comment by iklam** [comment]
Thanks @dholmes-ora and @tstuefe for the review.
/integrate
